src/translate.py

```diff
@@ -2934,10 +2967,60 @@ def array_access_from_add(
         index = addend
         scale = 1
 
+    if scale < 0:
+        scale = -scale
+        index = UnaryOp("-", index, type=index.type)
+
     target_type = base.type.get_pointer_target()
     if target_type is None:
         return None
 
+    uw_base = early_unwrap(base)
+    typepool = stack_info.global_info.typepool
+
+    # In `&x + index * scale`, if the type of `x` is not known, try to mark it as an array.
+    # Skip the `scale = 1` case because this often indicates a complex `index` expression,
+    # and is not actually a 1-byte array lookup.
+    if (
+        scale > 1
+        and offset == 0
+        and isinstance(uw_base, AddressOf)
+        and target_type.get_size_bytes() is None
+    ):
+        inner_type: Optional[Type] = None
+        if (
+            isinstance(uw_base.expr, GlobalSymbol)
+            and uw_base.expr.potential_array_dim(scale)[1] != 0
+        ):
+            # For GlobalSymbols, use the size of the asm data to check the feasibility of being
+            # an array with `scale`. This helps be more conservative around fake symbols.
+            pass
+        elif scale == 2:
+            # This *could* be a struct, but is much more likely to be an int
+            inner_type = Type.int_of_size(16)
+        elif scale == 4:
+            inner_type = Type.reg32(likely_float=False)
+        elif typepool.struct_field_inference and isinstance(uw_base.expr, GlobalSymbol):
+            # Make up a struct with a tag name based on the symbol & struct size.
+            struct_name = f"_struct_{uw_base.expr.symbol_name}_0x{scale:X}"
+            struct = typepool.get_struct_by_tag_name(
+                struct_name, stack_info.global_info.typemap
+            )
+            if struct is None:
+                struct = StructDeclaration.unknown_of_size(
+                    typepool, size=scale, tag_name=struct_name
+                )
+            elif struct.size != scale:
+                # This should only happen if there was already a struct with this name in the context
+                raise DecompFailure("sizeof(struct {struct_name}) != {scale:#x}")
+            inner_type = Type.struct(struct)
```

Comment:
This is the weird part mentioned in the description. I could take it out if it's *too* weird? It checks `typepool.struct_field_inference` so that it's only enabled when `--structs` is turned on.